### PR TITLE
Add slider bonus to Flashlight skill

### PR DESCRIPTION
**Pull request by MBmasher**
As of now, sliders are treated as hit circles in the Flashlight skill. However fast and long sliders introduce extra difficulty when playing a map with Flashlight.

This slider bonus comes in once the slider reaches a certain velocity (as slow sliders are easily readable). The bonus is also dependent on distance as long sliders require more memorisation to hit.

Therefore, most normal maps don't see a change, but an example of a map which gets buffed is [Kikoku Doukoku Jigokuraku [Notch Hell]](https://osu.ppy.sh/beatmapsets/437683#osu/942356) which goes from 356pp to 837pp for an SS with FL.

This change works independently, however if https://github.com/ppy/osu/pull/15728 and https://github.com/ppy/osu/pull/15728 are to be merged then `slider_multiplier` needs to be adjusted accordingly.

---

SR/PP sheet: https://docs.google.com/spreadsheets/d/1-tm0ybo7nUKbUDBSTlYbuJS-_o9Crg2xQbq9ik1icR4/edit#gid=1333939202

As of https://github.com/ppy/osu/pull/15666/commits/267d55a6a8f46598c4d7b016129459a92d09160e

### CODE REVIEW

**Comment by stanriders** [comment]
Honestly I'd add additional small bonus to all sliders due to dimming as well

**Comment by MBmasher** [comment]
> Honestly I'd add additional small bonus to all sliders due to dimming as well

I would say slider dimming makes no practical difference

**Comment by stanriders** [comment]
Was this adjusted after #15728 and #15728 merges? I'm also not entirely comfortable with Notch Hell being 800+ pp for SS, feels too much to me. What do other FL players think about it?

**Comment by smoogipoo** [comment]
What's the status of this PR? Do I need to make a sheet for it? Seems there's been discussion above without my posting of a sheet, have the concerns been addressed?

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs [hunk 1]
```diff
@@ -84,6 +87,19 @@ private double strainValueOf(DifficultyHitObject current)
             if (hidden)
                 result *= 1.0 + hidden_bonus;
 
+            double sliderBonus = 0.0;
+
+            if (osuCurrent.TravelTime != 0)
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs:92]: `if (osuCurrent.BaseObject is Slider)`

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs [hunk 2]
```diff
@@ -84,6 +87,19 @@ private double strainValueOf(DifficultyHitObject current)
...
+            {
+                // Reward sliders based on velocity.
+                double normalisedTravelDistance = osuCurrent.TravelDistance / scalingFactor;
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs:95]: Unless I'm missing something, `TravelDistance` is already being normalized in ODHO - https://github.com/ppy/osu/blob/7d93778355f7edb8d4c674f0707d69bcf5647c4e/osu.Game.Rulesets.Osu/Difficulty/Preprocessing/OsuDifficultyHitObject.cs#L213

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs [hunk 3]
```diff
@@ -84,6 +87,19 @@ private double strainValueOf(DifficultyHitObject current)
...
+                sliderBonus = Math.Pow(Math.Max(0.0, (normalisedTravelDistance) / osuCurrent.TravelTime - min_velocity), 0.5);
```
Comment (stanriders) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs:96]: `sliderBonus = Math.Pow(Math.Max(0.0, normalisedTravelDistance / osuCurrent.TravelTime - min_velocity), 0.5);`

**Review by stanriders — CHANGES_REQUESTED** [review]

**Comment by smoogipoo** [comment]
Spreadsheet: https://docs.google.com/spreadsheets/d/1T9iVUAfShfCtNn6wLzwgO6IlEBgtW38e3tleaOgI4cI/edit?usp=sharing

File: osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs (hunk 2 above)
Comment (smoogipoo) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/Flashlight.cs]: This was actually intentional as it's de-normalising the distance. I've refactored + documented better.

**Comment by smoogipoo** [comment]
@peppy

**Comment by smoogipoo** [comment]
@peppy

**Comment by stanriders** [comment]
There is still an issue with long repeated sliders as disscussed in pp dev so please do not merge.

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs [hunk 4]
```diff
@@ -73,6 +79,30 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool hidd
             if (hidden)
                 result *= 1.0 + hidden_bonus;
 
+            double sliderBonus = 0.0;
+
+            if (osuCurrent.BaseObject is Slider)
+            {
+                Debug.Assert(osuCurrent.TravelTime > 0);
+
+                Slider osuSlider = (Slider)(osuCurrent.BaseObject);
```
Comment (apollo-dw) [inline osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs:88]: You can save writing a line initializing `osuSlider` if you just do it in the if statement instead:
```csharp
if (osuCurrent.BaseObject is Slider osuSlider)
```

can do the same in `AimEvaluator.cs`

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs [hunk 5]
```diff
@@ -73,6 +79,30 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool hidd
...
+
+                // Reward sliders based on velocity.
+                sliderBonus = Math.Pow(Math.Max(0.0, pixelTravelDistance / osuCurrent.TravelTime - min_velocity), 0.5);
+
+                // Longer sliders require more memorisation.
+                sliderBonus *= pixelTravelDistance;
+
+                // Reward sliders with repeats.
+                if (osuSlider.RepeatCount > 0)
+                    sliderBonus *= repeat_bonus;
```
Comment (apollo-dw) [inline osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs:101]: A 3x repeat bonus, in Aim, is achieved with ~30 repeats. This immediately signals to me that this bonus is weighted too highly currently: either by `repeat_bonus` being too high, or by the full bonus being applied by only one repeat. Is this balanced around a specific map/case?

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs (hunk 4 above)
Comment (MBmasher) [reply osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs]: Fixed

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs [hunk 6]
```diff
@@ -15,11 +16,16 @@ public static class FlashlightEvaluator
...
 
+        private const double min_velocity = 0.5;
+        private const double slider_multiplier = 1.3;
+        private const double repeat_bonus = 3.0;
+
         /// <summary>
         /// Evaluates the difficulty of memorising and hitting an object, based on:
         /// <list type="bullet">
-        /// <item><description>distance between the previous and current object,</description></item>
+        /// <item><description>distance between previous objects and the current object,</description></item>
```
[the remark this answers is not in the document]
Comment (MBmasher) [reply osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs]: Fixed

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs [hunk 7]
```diff
@@ -73,6 +79,30 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool hidd
             if (hidden)
                 result *= 1.0 + hidden_bonus;
 
+            double sliderBonus = 0.0;
+
+            if (osuCurrent.BaseObject is Slider)
+            {
+                Debug.Assert(osuCurrent.TravelTime > 0);
```
[the remark this answers is not in the document]
Comment (MBmasher) [reply osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs]: Fixed

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs [hunk 8]
```diff
@@ -120,10 +121,15 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current, bool with
                 velocityChangeBonus *= Math.Pow(Math.Min(osuCurrObj.StrainTime, osuLastObj.StrainTime) / Math.Max(osuCurrObj.StrainTime, osuLastObj.StrainTime), 2);
             }
 
-            if (osuLastObj.TravelTime != 0)
+            if (osuLastObj.BaseObject is Slider)
             {
+                Debug.Assert(osuLastObj.TravelTime > 0);
```
[the remark this answers is not in the document]
Comment (MBmasher) [reply osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs]: Fixed

File: osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs (hunk 5 above)
Comment (MBmasher) [reply osu.Game.Rulesets.Osu/Difficulty/Evaluators/FlashlightEvaluator.cs]: So I was misunderstanding how `TravelDistance` worked. StanR originally pointed out the issue, linking me a map with a long slider with many repeats having a large spike in FL strain. I originally thought the issue was because of the newly added repeat bonus to `TravelDistance`, so I removed that and gave a flat bonus to all repeat sliders.

That wasn't the issue, rather the issue was that `TravelDistance` includes the distance traveled by repeats (which I didn't think was included).

I solved this by taking the slider bonus and dividing it by `Slider.RepeatCount + 1`, which should still give repeat sliders a bonus compared to normal sliders, but removes the possibility of abuse.

**Comment by apollo-dw** [comment]
Can we get an updated sheet for the latest changes? @smoogipoo

**Comment by smoogipoo** [comment]
SR/PP sheet added to description. It seems like the NM changes may be unexpected?

**Comment by apollo-dw** [comment]
Yeah, definitely as a result of moving the slider repeat bonus out of `OsuDifficultyHitObject`. Every single use of `TravelDistance` in `AimEvaluator` needs the slider repeat bonus applied as well if this is the direction. Might be easier to go with [my initial suggestion in pp dev](https://discord.com/channels/546120878908506119/549012787733594115/996010093659291698) of multiplying the repeat bonus to `TravelDistance` and not `LazyTravelDistance` in `OsuDifficultyHitObject`, and then using `LazyTravelDistance` in flashlight and applying your own repeat bonus there. Can give code example if too wordy.

For reference, this is an example of something missing the repeat bonus presently:
https://github.com/ppy/osu/blob/71912e10c4d31ed9b6a801c898224e974c2b776f/osu.Game.Rulesets.Osu/Difficulty/Evaluators/AimEvaluator.cs#L44

**Comment by MBmasher** [comment]
Went with @apollo-dw's suggestion to move repeat bonus to `TravelDistance`, this should be correct now

**Comment by smoogipoo** [comment]
Sheet updated, looks better.

**Review by apollo-dw — APPROVED** [review]

**Review by smoogipoo — APPROVED** [review]

**Comment by smoogipoo** [comment]
Awaiting 1 more review before merge from @ppy/osu-pp-committee

**Review by stanriders — APPROVED** [review]
